conda/gateways/repodata/__init__.py

```diff
@@ -289,3 +300,102 @@ def conda_http_errors(url, repodata_fn):
             e.response,
             caused_by=e,
         )
+
+
+class RepodataState(UserDict):
+    """
+    Load/save `.state.json` that accompanies cached `repodata.json`
+    """
+
+    _fields = (
+        "etag",
+        "mod",
+        "cache_control",
+        "size",
+    )
+
+    _aliased = ("_mod", "_etag", "_cache_control", "_url")
+
+    def __init__(self, cache_path_json, cache_path_state, repodata_fn):
+        super().__init__()
+        self.cache_path_json = pathlib.Path(cache_path_json)
+        self.cache_path_state = pathlib.Path(cache_path_state)
+        self.repodata_fn = repodata_fn
+
+    def load(self):
+        """
+        Cache headers and additional data needed to keep track of the cache are
+        stored separately, instead of the previous "added to repodata.json"
+        arrangement.
+        """
+        try:
+            state_path = self.cache_path_state
+            log.debug("Load %s cache from %s", self.repodata_fn, state_path)
+            state = json.loads(state_path.read_text())
+            # json and state files should match
+            json_stat = self.cache_path_json.stat()
+            if not (
+                state.get("mtime_ns") == json_stat.st_mtime_ns
+                and state.get("size") == json_stat.st_size
```

Comment:
Are we testing the cache invalidation flow thoroughly enough, to verify that it does or does not use the cache depending, and that it doesn't think that a concurrent conda overwrote repodata.json while we are trying to overwrite repodata.state.json?